### PR TITLE
[SPARK-33827][SS] Unload inactive state store as soon as possible

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to unload inactive state store as soon as possible. The timing of unload inactive state stores, happens when we get to load active state store provider at executors. At the time, state store coordinator will return back the state store provider list including loaded stores that are already loaded by other executors in new batch. Each state store provider in the list will go to unload.

### Why are the changes needed?

Per the discussion at #30770, it makes sense to me we should unload inactive state store asap. Now we run a maintenance task periodically to unload inactive state stores. So there will be some delays between a state store becomes inactive and it is unloaded.

However, we can force Spark to always allocate a state store to same executor, by using task locality configuration. This can reduce the possibility to have inactive state store.

Normally, with locality configuration, we might not able to see inactive state store generally. There is still chance an executor can be failed and reallocated, but in this case, inactive state store is also lost too. So it is not an issue.

Making driver-executor bi-directional for unloading inactive state store looks non-trivial, and seems to me, it is not worth, after considering what we can do with locality.

This proposes a simpler but effective approach. We can check if loaded state store is already loaded at other executor during reporting active state store to the coordinator. If so, it means the loaded store is inactive now, and it is going to be unload by the next maintenance task. Then we unload that store immediately.

How do we make sure the loaded state store in previous batch is loaded at other executor in this batch before reporting in this executor? With task locality and preferred location, once an executor is ready to be scheduled, Spark should assign the state store provider previously loaded at the executor. So when this executor gets a new assignment other than previously loaded state store, it means the previously loaded one is already assigned to other executor.

There is still a delay between the state store is loaded at other executor, and unloading it when reporting active state store at this executor. But it should be minimized now. And there won't be multiple state store belonging to same operator are loaded at the same time at one single executor, because once the executor reports any active store, it will unload all inactive stores. This should not be an issue IMHO.

This is a minimal change to unload inactive state store asap without significant change.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HeartSaVioR

**Comment by viirya** [comment]
cc @tdas @zsxwing @jose-torres @gaborgsomogyi @xuanyuanking too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala [hunk 1]
```diff
@@ -122,13 +125,20 @@ private class StateStoreCoordinator(override val rpcEnv: RpcEnv)
     extends ThreadSafeRpcEndpoint with Logging {
   private val instances = new mutable.HashMap[StateStoreProviderId, ExecutorCacheTaskLocation]
 
-  override def receive: PartialFunction[Any, Unit] = {
-    case ReportActiveInstance(id, host, executorId) =>
+  override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case ReportActiveInstance(id, host, executorId, otherProviderIds) =>
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala:129]: Given the semantic of the request, I'm feeling more natural for this request to return provider ids which are assigned to this `ExecutorCacheTaskLocation`, say, "Here's the state store IDs you're registered as active.". That makes sure the request does the only a thing like we do for function - this change seems to do two different semantics of requests at once, and now return value doesn't look to be related to the origin semantic of request.

But I agree this change is better in point of performance-wise, as this doesn't require iterating on instances. Let's hear how others think about this.

**Comment by HeartSaVioR** [comment]
The direction looks OK to me. Let's wait a bit for others' voices on the input/output of `ReportActiveInstance`. Thanks for the change!

**Review by xuanyuanking — APPROVED** [review]
LGTM
Only one concern on the RPC change, but I don't think it's a big issue now. After this change, the request body contains duplicated messages(e.g., every state store id contains `checkpointRootLocation`), but the RPC is a low-frequency call.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala:129]: I'm OK to document ReportActiveInstance to describe the behavior. Probably changing the last parameter name to `providerIdsToCheckActive` (or `providerIdsToCheck` for shorter) would help understanding the behavior as well.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala]: Okay. I added a few comment and change the parameter name.

**Review by HeartSaVioR — APPROVED** [review]
+1

I'll merge in early next week once there's no further comment.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Github Action passed. Merging.

**Comment by HeartSaVioR** [comment]
Thanks for the contribution! Merged into master.

**Comment by viirya** [comment]
Thanks guys!

**Pull request closed without merging**